### PR TITLE
Reactivate EMAC Greentea test

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

This PR reactivates the emac test (under connectivity/netsocket/tests/TESTS/network/emac).  This test is a bit complex because it works by communicating with a second board running a CTP sever application.  To avoid build system cancer, I had to move this application into [its own repository](https://github.com/mbed-ce/mbed-ethernet-ctp-server).  I also moved the common code used by both the CTP server and the Mbed test into its own CMake library, so they can share like 98% of their codebase.

Happily, once I did all this, the test worked fine on both devices I tested.

#### Impact of changes 
- EMAC test works again, so we have a way to verify that ethernet works

### Documentation 

Updated https://github.com/mbed-ce/mbed-os/wiki/Running-the-Mbed-Greentea-Tests-Locally

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR
    
I verified that the EMAC test passes on both LPC1768 and NUCLEO_F429ZI, using a K64F board as the CTP tester.

----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: .github/workflows/basic_checks.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ jobs:
 
     steps:
       - name: Checkout repo
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
```
Comment (multiplemonomials) [inline .github/workflows/basic_checks.yml:20]: Updating these as they are deprecated and producing warnings in the Github Actions logs

File: README.md [hunk 2]
```diff
@@ -13,8 +13,9 @@ Arm Mbed OS is an open source embedded operating system designed specifically fo
 Mbed OS provides a platform that includes:
 
 - Security foundations.
-- Cloud management services.
-- Drivers for sensors, I/O devices and connectivity.
+- Embedded networking libraries, e.g Wi-Fi and Ethernet drivers and network libraries using them
```
Comment (multiplemonomials) [inline README.md:16]: Decided it was time to update this readme, as Mbed CE doesn't actually provide any cloud management services

File: connectivity/libraries/nanostack-libservice/CMakeLists.txt [hunk 3]
```diff
@@ -22,6 +22,8 @@ target_sources(mbed-nanostack-libservice
         source/nvmHelper/ns_nvm_helper.c
 )
 
+target_link_libraries(mbed-nanostack-libservice PUBLIC mbed-nanostack-hal_mbed_cmsis_rtos)
```
Comment (multiplemonomials) [inline connectivity/libraries/nanostack-libservice/CMakeLists.txt:25]: This dependency had been missing, no idea how it worked until now

File: connectivity/netsocket/tests/TESTS/network/emac/emac_test_memory.cpp [hunk 4]
```diff
@@ -152,10 +152,10 @@ void test_emac_memory_cb(int opt)
 void test_emac_memory()
 {
     RESET_ALL_ERROR_FLAGS;
-    SET_TRACE_LEVEL(TRACE_SEND | TRACE_SUCCESS | TRACE_FAILURE);
+    SET_TRACE_LEVEL(TRACE_FAILURE);
```
Comment (multiplemonomials) [inline connectivity/netsocket/tests/TESTS/network/emac/emac_test_memory.cpp:155]: This test produced a TON of logs so I turned down the verbosity

File: connectivity/netsocket/tests/emac_test_utils/EmacTestNetworkStack.h [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// Get a pointer to the EMAC driver.  Will return nullptr if no interface has been added yet.
+    EMAC *get_emac();
+
+    /// Get the MAC address being used for the ethernet port
+    unsigned char const *get_mac_addr() const;
+
+protected:
+
+    nsapi_error_t socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto) override;
```
Comment (multiplemonomials) [inline connectivity/netsocket/tests/emac_test_utils/EmacTestNetworkStack.h:92]: I migrated this header from emac_TestNetworkStack.h.  It had a lot of issues; namely it didn't use any `override` declarations so a lot of functions had changed signature and weren't actually overriding anymore.  I removed the nonessential ones and added override to the rest.

File: run-formatter.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# Script to run the Astyle formatter on the Mbed OS code base.
```
Comment (multiplemonomials) [inline run-formatter.sh:3]: Got tired of doing this by hand so I added this script!

File: tools/cmake/UploadMethodManager.cmake [hunk 7]
```diff
@@ -66,11 +66,12 @@ set(MBED_GDB_PORT 23331 CACHE STRING "Port that the GDB server will be started o
 set(MBED_UPLOAD_SERIAL_NUMBER "" CACHE STRING "Serial number of the Mbed board or the programming tool, for upload methods that select by serial number.")
 
 # Handle legacy per-upload-method aliases for the upload serial number
-foreach(LEGACY_VAR_NAME JLINK_USB_SERIAL_NUMBER LINKSERVER_PROBE_SN MBED_TARGET_UID OPENOCD_ADAPTER_SERIAL PYOCD_PROBE_UID STLINK_SERIAL_ARGUMENT STM32CUBE_PROBE_SN)
+foreach(LEGACY_VAR_NAME JLINK_USB_SERIAL_NUMBER LINKSERVER_PROBE_SN MBED_TARGET_UID OPENOCD_ADAPTER_SERIAL PYOCD_PROBE_UID STLINK_PROBE_SN STM32CUBE_PROBE_SN)
```
Comment (multiplemonomials) [inline tools/cmake/UploadMethodManager.cmake:69]: Oops there was a bit of a bug here, this loop had the wrong variable name so it was messing with an internal variable used by the STLINK upload method!  Caused it to generate invalid arguments for the st-flash tool.

File: tools/python/memap/memap.py [hunk 8]
```diff
@@ -99,7 +99,7 @@ def contains_addr(self, addr: int) -> bool:
 
 class _Parser(ABC):
     """Internal interface for parsing"""
-    SECTIONS = ('.text', '.data', '.bss', '.heap', '.stack')
+    SECTIONS = ('.text', '.data', '.bss', '.heap', '.heap_0', '.stack')
```
Comment (multiplemonomials) [inline tools/python/memap/memap.py:102]: Fixing a memap bug -- for targets with split heap support, it was not accepting heap_0 as a valid section name, so the heap memory wouldn't be special-cased correctly and would get treated like it was 100% full of data.

**Review by JohnK1987 — APPROVED** [review]

File: connectivity/netsocket/tests/TESTS/network/emac/emac_test_memory.cpp (hunk 4 above)
Comment (multiplemonomials) [reply connectivity/netsocket/tests/TESTS/network/emac/emac_test_memory.cpp:155]: Ah, it says in the README for the test how to adjust this verbosity.  This was how I figured out how to change it!

**Pull request merged**